### PR TITLE
Enable DIP25 by default

### PR DESCRIPTION
**Pull request by Geod24**
Prior art: https://github.com/dlang/dmd/pull/9154
~Just want to see what breaks (loads, probably).~

### CODE REVIEW

**Comment by Geod24** [comment]
No longer a WIP

File: src/dmd/escape.d [hunk 1]
```diff
@@ -1042,13 +1042,19 @@ bool checkNewEscape(Scope* sc, Expression e, bool gag)
...
             if (!gag)
             {
                 const(char)* kind = (v.storage_class & STC.parameter) ? "parameter" : "local";
-                error(e.loc, "copying `%s` into allocated memory escapes a reference to %s variable `%s`",
-                    e.toChars(), kind, v.toChars());
+                if (emitError)
+                    error(e.loc, "copying `%s` into allocated memory escapes a reference to %s variable `%s`",
+                          e.toChars(), kind, v.toChars());
+                else
+                    error(e.loc, "copying `%s` into allocated memory escapes a reference to %s variable `%s`",
```
Comment (thewilsonator) [inline src/dmd/escape.d]: Second one should be a deprecation. Please factorise the message out.

File: src/dmd/escape.d [hunk 2]
```diff
@@ -1334,8 +1351,10 @@ private bool checkReturnEscapeImpl(Scope* sc, Expression e, bool refs, bool gag)
                     TypeFunction tf = cast(TypeFunction)fd.type;
                     if (tf.isref)
                     {
-                        if (!gag)
+                        if (!gag && emitError)
                             error(e.loc, "escaping reference to outer local variable `%s`", v.toChars());
+                        else if (!gag)
+                            deprecation(e.loc, "escaping reference to outer local variable `%s`", v.toChars());
```
Comment (thewilsonator) [inline src/dmd/escape.d]: Please factorise the message.

**Comment by Geod24** [comment]
Addressed @thewilsonator 's review.

**Review by thewilsonator — APPROVED** [review]

**Comment by Geod24** [comment]
I tried to cover the other code path, which seems uncovered, but ended up finding unsafe escapes...

```D
struct Data
{
    @safe @nogc:
    char[256] buffer;

    void moreEscape(const(char)[][] ptr) const pure nothrow
    {
        ptr[0] = this.buffer[];
    }
}

void main() @safe @nogc
{
    const value = foo();
    assert(value !is null);
}

const(char)[] foo () @safe @nogc
{
    Data d;
    const(char)[][1] ptr;
    d.moreEscape(ptr);
    assert(ptr[0] !is null);
    return ptr[0];
}
```

Compiles just fine with `-preview=dip25` and `-preview=dip1000`.
I find it very similar to https://issues.dlang.org/show_bug.cgi?id=20505 .
@WalterBright ?

**Comment by Geod24** [comment]
The error message does trigger on some code in swarm, dlang-tour/core, and ae.
I've pushed a tentative fix for the failures. I can see the error message in the testsuite, but all the tests I've found have `-preview=dip1000` enabled.

File: src/dmd/escape.d [hunk 3]
```diff
@@ -1042,15 +1042,20 @@ bool checkNewEscape(Scope* sc, Expression e, bool gag)
...
-                error(e.loc, "copying `%s` into allocated memory escapes a reference to %s variable `%s`",
-                    e.toChars(), kind, v.toChars());
+                const(char)* msg = "copying `%s` into allocated memory escapes a reference to %s variable `%s`";
+                if (emitError)
+                    error(e.loc, msg, e.toChars(), kind, v.toChars());
+                else
+                    deprecation(e.loc, msg, e.toChars(), kind, v.toChars());
             }
-            result = true;
+            result = result || emitError;
```
Comment (WalterBright) [inline src/dmd/escape.d]: `result |= emitError;`

**Comment by WalterBright** [comment]
I don't understand. Currently, it is `-preview=dip25`. Shouldn't it be a simple change to `-revert=dip25`? Why all the code changes?

**Comment by Geod24** [comment]
Because `-preview` switch are great for providing our users with a preview of a new feature, gauging interest in it, allowing iterative design and implementation.

However, their opt-in nature means they cannot and should not be used as a replacement to a proper deprecation cycle.

When I tried to enable the preview by default, half the projects on Buildkite broke. I went ahead and fixed some of them (or asked for releases for the projects that had been fixed already), but that is a very good indication why we shouldn't turn on preview switches by default.

**Comment by WalterBright** [comment]
> However, their opt-in nature means they cannot and should not be used as a replacement to a proper deprecation cycle.

We switched to the preview/revert system because the deprecation cycle was not working. All the users have to do to "fix" their code is add the revert switch to their compiles. It could hardly be easier.

People just ignore the deprecation messages. For proof, just look at the logs building phobos, druntime, test files, etc. Deprecation messages that nobody fixes.

File: src/dmd/escape.d [hunk 4]
```diff
@@ -1072,17 +1077,19 @@ bool checkNewEscape(Scope* sc, Expression e, bool gag)
...
+                const emitError = global.params.useDIP25;
                 // https://dlang.org/spec/function.html#return-ref-parameters
                 // Only look for errors if in module listed on command line
-
                 if (p == sc.func)
                 {
                     //printf("escaping reference to local ref variable %s\n", v.toChars());
                     //printf("storage class = x%llx\n", v.storage_class);
-                    escapingRef(v);
+                    escapingRef(v, emitError);
```
Comment (WalterBright) [inline src/dmd/escape.d]: It looks like all `emitError` is is a copy of `global.params.useDIP25`. The complexity of this change would be significantly reduced if you just used the global. The global is a hack anyway that will be eventually removed anyway.

Comment (WalterBright) [reply src/dmd/escape.d]: If you must use a parameter, just call it `useDIP25`, then nobody has to trace through to see that that is the only setting it is ever set to.

File: src/dmd/escape.d [hunk 5]
```diff
@@ -1092,10 +1099,12 @@ bool checkNewEscape(Scope* sc, Expression e, bool gag)
...
                     {
-                        if (!gag)
-                            error(e.loc, "storing reference to outer local variable `%s` into allocated memory causes it to escape",
-                                  v.toChars());
-                        result = true;
+                        const(char)* msg = "storing reference to outer local variable `%s` into allocated memory causes it to escape";
+                        if (!gag && emitError)
+                            error(e.loc, msg, v.toChars());
+                        else if (!gag)
+                            deprecation(e.loc, msg, v.toChars());
```
Comment (WalterBright) [inline src/dmd/escape.d]: Notice the arguments to `error` and `deprecation` are exactly the same. This suggests that `emitError` could be replaced with a delegate (or function pointer) that is set to `error` or `deprecation` and passed instead. This pattern repeats 4 times.

**Comment by wilzbach** [comment]
> We switched to the preview/revert system because the deprecation cycle was not working. All the users have to do to "fix" their code is add the revert switch to their compiles. It could hardly be easier.

Uhm I made these preview/revert PRs because we introduced a new flag for every new experimental feature, which didn't scale very well.

The problem is that people have dependencies in their code that they don't control. Without a deprecation they can't compile their project anymore. With deprecations they can bug the maintainers of their dependencies (and those maintainers can fix them), but are still able to use the latest/greatest D frontend.

tl;dr: deprecations are awesome and an absolute must for a stable ecosystem (even if it doesn't look like people are fixing them)

>  For proof, just look at the logs building phobos, druntime, test files, etc. Deprecation messages that nobody fixes.

Druntime and Phobos are built with `-de` for that very reason ;-)

**Comment by Geod24** [comment]
> We switched to the preview/revert system because the deprecation cycle was not working. All the users have to do to "fix" their code is add the revert switch to their compiles. It could hardly be easier.

I'm sorry but that is simply not true. Deprecation are working very well. One could argue there are a few adjustments needed in our process, but they have been doing an awesome job.

How many things did we deprecate over the last few years ? Dozens, if not hundreds. We used them generously to get rid of unwanted/duplicated/obsolete features (hexstring, delete operator, allocator/deallocator, D1 operators overload...), and more importantly, constructs that could be dangerous but could also be used correctly or would lead to bugs only in rare case (e.g. using the return of a coma expression, some shadowing cases, initialization of `immutable` from `static this`...)

Trying to do the same with `-preview`/`-revert` switch would never scale. It would also create a terrible user experience.
DIP25 has been added in [v2.067.0](https://dlang.org/changelog/2.067.0.html), a good 5 years ago. As mentioned, half of Buildkite went red when I enabled it. In contrast, one of the largest language deprecation we'd ever done, [changing the behavior of imports](https://dlang.org/changelog/2.071.0.html), was done more than a year later via deprecations, and everything has long been fixed.
How many `-preview` have made it to be the default ? Exactly 0, this would be the first one.

> People just ignore the deprecation messages. For proof, just look at the logs building phobos, druntime, test files, etc. Deprecation messages that nobody fixes.

Once again, not true at all. Those projects compile with `-de`. Most projects clean up deprecations when they upgrade compiler. I don't think I've ever seen a project with `-d` enabled, but I've seen quite a few with `-de`.

On the other hand, how many use `-preview=dip{25,1000,1008}` ?
It took *two and a half years* to get Phobos to compile with DIP1000... The compiler support was introduced in 2.073.0, January 2017, and Phobos support came in [v2.087.0](https://dlang.org/changelog/2.087.0.html#phobos-dip1000), July 2019.

This is not because `-preview` is a bad replacement for deprecations, but because features under `-preview` are incomplete and subject to change (even design changes). We still have quite a few issues with DIP1000 (bugzilla yields 57 issues on "DIP1000") but at least people can try it out in their projects.

The problem at hands is that features in `-preview` never seem to be completed to the point they can turn into a deprecation. They keep piling up, but as mentioned, nothing made it past an opt-in switch.

**Comment by 12345swordy** [comment]
Aren't we merging this or not?

File: src/dmd/escape.d (hunk 4 above)
Comment (Geod24) [reply src/dmd/escape.d]: The issue is that `escapingRef` is not only called for DIP25.
For example here, it is also called at L1070

File: src/dmd/escape.d (hunk 5 above)
Comment (Geod24) [reply src/dmd/escape.d]: The delegate passed would also need to handle setting the `result` variable, which would be more involved than just hoisting the parameter.
The repetition is also due to the code being itself repetitive. I originally started to refactor it, but the diff was just too much.

**Comment by Geod24** [comment]
Addressed the `|=` review and answered comments.

**Comment by thewilsonator** [comment]
is this good to go?

**Review by RazvanN7 — APPROVED** [review]

**Comment by WalterBright** [comment]
> Once again, not true at all.

Then things have changed fairly recently, as I regularly recall Phobos emitting deprecation messages that went on for release after release. I'm glad it has changed.
